### PR TITLE
Add generic unsafeAllow option and granular NatSpec overrides

### LINKED ISSUE #280: Add a manual override for every check
Requested in https://forum.openzeppelin.com/t/force-override-upgrade-validation/5368:

>When attempting to use openzeppelin-upgrades we are able to override only a particular subset of unsafe actions (namely unsafeAllowCustomTypes and unsafeAllowLinkedLibraries). However in some cases, one may want to override things like “Use of delegatecall is not allowed” (e.g. when delegate calls are only used to read contract storage).
>
>Would it be within the realm of possibility to include a manual override for all validations, or simply a forceSkipValidation feature?

### LINKED ISSUE #318: Granular "unsafeAllow" through NatSpec comments
The latest Solidity allows adding NatSpec comments to any statement. We can use this to disable or enable checks on a more granular level than we do now.

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #318 
Fixes #280

Changelog
===

Core - Public
---

- Add a new field to the ValidationOptions interface unsafeAllow with can be used to silent any error type. For example, opts = { unsafeAllow: ['external-library-linking', 'delegatecall'] } will silent the corresponding checks.
- Custom natspec comments can disable check errors directly from the solidity code. See core/contracts/test/ValidationNatspec.sol for usage of these natspec comments.

Core - Internal
---
- Add a new error type 'inline-assembly'
- Add check for inline blocks in the ast (disabled for now)
- Fix a bug with library placeholder when generating smart contract versions.

### CODE REVIEW

File: packages/core/src/validate/run.ts [hunk 1]
```diff
@@ -120,25 +175,32 @@ function* getConstructorErrors(contractDef: ContractDefinition, decodeSrc: SrcDe
...
       };
     }
   }
+  /// TODO: uncomment to enable assembly check
+  // for (const node of findAll('InlineAssembly', contractDef, node => skipCheck('inline-assembly', node))) {
+  //   yield {
+  //     kind: 'inline-assembly',
+  //     src: decodeSrc(node),
+  //   };
+  // }
```
Comment (frangio) [inline packages/core/src/validate/run.ts:203]: Comment should be removed.

File: packages/core/contracts/test/Validations.sol [hunk 2]
```diff
@@ -160,3 +160,11 @@ contract UsesExplicitUnsafeExternalLibrary {
     UnsafeExternalLibrary.explode(x, y);
   }
 }
+
+contract HasInlineAssembly {
```
Comment (frangio) [inline packages/core/contracts/test/Validations.sol:164]: More inline assembly stuff that should be removed I think.

File: packages/core/contracts/test/ValidationsNatspec.sol [hunk 3]
```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.2;
+
+contract HasEmptyConstructorNatspec {
+  constructor() { }
+}
```
Comment (frangio) [inline packages/core/contracts/test/ValidationsNatspec.sol:6]: This one doesn't need any special comments so shouldn't be needed right?

File: packages/core/contracts/test/ValidationsNatspec.sol [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+/// @custom:openzeppelin-upgrade-allow-unsafe state-variable-assignment
+contract HasStateVariableAssignmentNatspec1 {
+  uint x = 1;
+}
+
+contract HasStateVariableAssignmentNatspec2 {
+  /// @custom:openzeppelin-upgrade-allow-unsafe state-variable-assignment
+  uint x = 1;
+}
```
Comment (frangio) [inline packages/core/contracts/test/ValidationsNatspec.sol:31]: I'd like to see a test that something like the following fails:

```solidity
contract HasStateVariableAssignmentNatspec2 {
  /// @custom:openzeppelin-upgrade-allow-unsafe state-variable-assignment
  uint x = 1;

  uint y = 2;
}
```

Consider adding a few more tests like this that should fail to validate, for other overrides. No need to be exhaustive, just a few strategically chosen ones.

File: packages/core/package.json [hunk 5]
```diff
@@ -26,7 +26,7 @@
   "devDependencies": {
     "@ava/typescript": "^1.1.1",
     "@nomiclabs/hardhat-etherscan": "^2.1.0",
-    "@openzeppelin/contracts": "^3.1.0",
+    "@openzeppelin/contracts": "^4.0.0",
```
Comment (frangio) [inline packages/core/package.json:29]: This looks unrelated to this PR, along with all of the changes to the contracts in `core`.

File: packages/core/src/validate/overrides.ts [hunk 6]
```diff
@@ -5,67 +5,68 @@ import { ValidationError } from './run';
...
-  contractName: string,
-  errorsToProcess: ValidationError[],
-  opts: ValidationOptions,
-): ValidationError[] {
-  const { unsafeAllowCustomTypes, unsafeAllowLinkedLibraries } = withValidationDefaults(opts);
-  let errors: ValidationError[] = errorsToProcess;
+export function withValidationDefaults(opts: ValidationOptions): Required<ValidationOptions> {
+  const unsafeAllowCustomTypes = opts.unsafeAllowCustomTypes ?? false;
+  const unsafeAllowLinkedLibraries = opts.unsafeAllowLinkedLibraries ?? false;
+  const unsafeAllow = opts.unsafeAllow ?? [];
```
Comment (frangio) [inline packages/core/src/validate/overrides.ts:32]: I think this function should always return options that are consistent: `unsafeAllowCustomTypes` should be `true` if and only if `unsafeAllow.includes('struct-definition')` and `unsafeAllow.includes('enum-definition')`. Right now it's only ensuring a single direction (left to right).

An easy way to ensure the other direction is to set the initial values based on the array:
```suggestion
  const unsafeAllow = opts.unsafeAllow ?? [];
  const unsafeAllowCustomTypes =
    opts.unsafeAllowCustomTypes ||
    (unsafeAllow.includes('struct-definition') &&
      unsafeAllow.includes('enum-definition'));
  const unsafeAllowLinkedLibraries =
    opts.unsafeAllowLinkedLibraries ||
    unsafeAllow.includes('external-library-linking');
```

Just a suggestion, feel free to implement in a different way.

File: packages/core/src/version.ts [hunk 7]
```diff
@@ -20,6 +20,15 @@ export function getVersion(bytecode: string, linkedBytecode?: string): Version {
 }
 
 export function hashBytecode(bytecode: string): string {
```
Comment (frangio) [inline packages/core/src/version.ts:22]: Although I think we should ship this fix as soon as possible, can we remove it from this PR so we keep this one about overrides only?

File: packages/core/src/validate/overrides.ts [hunk 8]
```diff
@@ -5,67 +5,68 @@ import { ValidationError } from './run';
...
-    );
+      errors = errors.filter(error => {
+        const isException = errorType === error.kind;
+        exceptionsFound = exceptionsFound || isException;
+        return !isException;
+      });
+
+      if (exceptionsFound && !silenced && errorDescription) {
+        console.error(
+          '\n' +
```
Comment (frangio) [inline packages/core/src/validate/overrides.ts:63]: I think I've removed these initial newlines before, I think it looks better without them.

**Review by frangio — COMMENTED** [review]
Awesome stuff!

Please remove everything related to inline assembly (lots of things!), which I think we agreed to do later.

File: packages/core/src/version.ts (hunk 7 above)
Comment (Amxx) [reply packages/core/src/version.ts:22]: The fix is needed for the tests to work correctly

**Review by frangio — APPROVED** [review]
LGTM!

**Comment by frangio** [comment]
I remembered one last thing I wanted us to change. Let's make the NatSpec tag into `@custom:oz-upgrades-unsafe-allow`. (Note that "upgrades" is plural.)

And I think this is missing documentation as well. We need to be clear this only works with Solidity 0.8.

**Review by frangio — CHANGES_REQUESTED** [review]
I remembered one last thing I wanted us to change. Let's make the NatSpec tag into @custom:oz-upgrades-unsafe-allow. (Note that "upgrades" is plural.)

And I think this is missing documentation as well. We need to be clear this only works with Solidity 0.8.

**Comment by frangio** [comment]
There is a `CHANGELOG.md` file in each package. You can add a new `## Unreleased` section at the top with the changes introduced in this PR.

With changelog and documentation this is good to merge. After merging we should write up some more details about the bug that was fixed and assess whether anyone could have been affected by this and what they should do. It's a little complicated to think through this, we can look at it together on Friday.

**Comment by Amxx** [comment]
Changelog and documentation added

**Review by frangio — APPROVED** [review]
LGTM!

**Comment by bh2smith** [comment]
Awesome work here! Thanks for the changes, now this library is a whole lot more accessible!
